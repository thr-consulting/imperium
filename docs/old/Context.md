---
id: context
title: Context
sidebar_label: Context
---

A Context object instance provides access to all the app's models, authentication information and connectors.

## Methods

### `addModule(moduleFunction)`
Adds a [server module](ServerModule.md) function to the Context instance.

### `getModel(name)`
Returns a specific model.

### `get models()`
Returns an object with all the models.

### `set auth(value)`
Sets the authentication information for this Context instance.

### `get auth()`
Gets the authentication information.

### `get connectors()`
Gets the connectors object.
