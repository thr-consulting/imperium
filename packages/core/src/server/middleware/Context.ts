import {Connectors, ServerModuleFunc} from '../../../types';

export default class Context {
	_connectors: Connectors[];
	_models: Record<string, any>;
	_auth: any;

	constructor(connectors) {
		this._connectors = connectors;
		this._models = {};
		this._auth = null;
	}

	/**
	 * Adds a module function to the registry. A module function has the
	 * following signature: (connectors, context) => {} and returns
	 * an object keyed with data model objects.
	 * @param moduleFunc
	 */
	addModule(moduleFunc: ServerModuleFunc): void {
		const moduleModels = moduleFunc(this._connectors, this);
		this._models = Object.assign({}, this._models, moduleModels);
	}

	/**
	 * Returns the model with the specified name
	 * @param name
	 * @returns {*}
	 */
	getModel(name): any {
		return this._models[name];
	}

	/**
	 * Returns all the models
	 * @returns {*}
	 */
	get models(): Record<string, any> {
		return this._models;
	}

	/**
	 * Sets new auth data
	 * @param value
	 */
	set auth(value) {
		this._auth = value;
	}

	/**
	 * Returns the auth data
	 * @returns {null|*}
	 */
	get auth(): any {
		return this._auth;
	}

	get connectors(): Connectors[] {
		return this._connectors;
	}
}
