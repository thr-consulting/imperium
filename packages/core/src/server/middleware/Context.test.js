import Context from './Context';

const myConnectors = {
	MyConnector: {},
};

function moduleFunc(/* connectors, context */) {
	return {
		MyDataModel: {
			fakeApiFn: () => 'fake api result',
		},
	};
}

describe('Context', () => {
	it('should add modules properly', () => {
		const context = new Context(myConnectors);
		context.addModule(moduleFunc);
		expect(context.getModel('MyDataModel')).toMatchSnapshot();
		expect(context.models.MyDataModel).toMatchSnapshot();
	});

	it('should set and get auth data', () => {
		const context = new Context(myConnectors);
		context.auth = {name: 'test'};
		expect(context.auth).toMatchSnapshot();
	});

	it('should retrieve the connectors', () => {
		const context = new Context(myConnectors);
		expect(context.connectors).toMatchSnapshot();
	});
});
