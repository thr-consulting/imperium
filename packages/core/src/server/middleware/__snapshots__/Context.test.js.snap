// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Context should add modules properly 1`] = `
Object {
  "fakeApiFn": [Function],
}
`;

exports[`Context should add modules properly 2`] = `
Object {
  "fakeApiFn": [Function],
}
`;

exports[`Context should retrieve the connectors 1`] = `
Object {
  "MyConnector": Object {},
}
`;

exports[`Context should set and get auth data 1`] = `
Object {
  "name": "test",
}
`;
